dali/kernels/imgproc/sampler.h

```diff
@@ -30,165 +30,183 @@ namespace kernels {
+}
+
+template <int _spatial_ndim, typename In>
+struct Sampler<DALI_INTERP_NN, _spatial_ndim, In> {
+  static constexpr int spatial_ndim = _spatial_ndim;
+
+  using icoords = dali::ivec<spatial_ndim>;
+  using fcoords = dali::vec<spatial_ndim>;
+
   Sampler() = default;
-  DALI_HOST_DEV explicit Sampler(const Surface2D<const In> &surface)
+  DALI_HOST_DEV explicit Sampler(const Surface<spatial_ndim, const In> &surface)
   : surface(surface) {}
 
-  Surface2D<const In> surface;
+  Surface<spatial_ndim, const In> surface;
 
   template <typename T = In, typename BorderValue>
-  DALI_HOST_DEV T at(
-      int x, int y, int c,
-      BorderValue border_value) const {
-    if (x < 0 || x >= surface.size.x ||
-        y < 0 || y >= surface.size.y) {
-      return ConvertSat<T>(GetBorderChannel(border_value, c));
+  DALI_HOST_DEV DALI_FORCEINLINE T at(icoords pos, int c, BorderValue border_value) const {
+    if (all_in_range(pos, surface.size)) {
+      return ConvertSat<T>(surface(pos, c));
     } else {
-      return ConvertSat<T>(surface(x, y, c));
+      return ConvertSat<T>(GetBorderChannel(border_value, c));
     }
   }
 
   template <typename T = In>
-  DALI_HOST_DEV T at(
-      int x, int y, int c,
-      BorderClamp) const {
-    ivec2 clamped = clamp(ivec2(x, y), ivec2(0, 0), surface.size - 1);
-    return ConvertSat<T>(surface(clamped, c));
-  }
-
-  template <typename T = In, typename BorderValue>
-  DALI_HOST_DEV T at(
-      float x, float y, int c,
-      BorderValue border_value) const {
-    return at<T>(floor_int(x), floor_int(y), c, border_value);
+  DALI_HOST_DEV DALI_FORCEINLINE T at(icoords pos, int c, BorderClamp) const {
+    return ConvertSat<T>(surface(clamp(pos, icoords(0), surface.size - 1), c));
   }
 
   template <typename T = In, typename BorderValue>
-  DALI_HOST_DEV T at(ivec2 xy, int c, BorderValue border_value) {
-    return at(xy.x, xy.y, c, border_value);
-  }
-
-  template <typename T = In, typename BorderValue>
-  DALI_HOST_DEV T at(vec2 xy, int c, BorderValue border_value) {
-    return at(xy.x, xy.y, c, border_value);
-  }
-
-  template <typename T, typename BorderValue>
-  DALI_HOST_DEV void
-  operator()(T *pixel, int x, int y, int c, BorderValue border_value) const {
-    pixel[c] = at<T>(x, y, c, border_value);
-  }
-
-  template <typename T, typename BorderValue>
-  DALI_HOST_DEV void
-  operator()(T *pixel, float x, float y, int c, BorderValue border_value) const {
-    pixel[c] = at<T>(x, y, c, border_value);
+  DALI_HOST_DEV DALI_FORCEINLINE T at(fcoords pos, int c, BorderValue border_value) const {
+    return at<T>(floor_int(pos), c, border_value);
   }
 
   template <typename T, typename BorderValue>
   DALI_HOST_DEV void operator()(
       T *pixel,
-      int x, int y,
+      icoords pos,
       BorderValue border_value) const {
-    if (x < 0 || x >= surface.size.x ||
-        y < 0 || y >= surface.size.y) {
+    if (all_in_range(pos, surface.size)) {
```

Comment:
Why can't you just call at(...):
```
pixel[c] = at(pos, c , border_value)
```

Reply:
Or it can use `void operator()(T *pixel, icoords pos, int c, BorderValue border_value) `.

Reply:
I wanted to move the range check outside the loop over channels. Compilers are smart, but they do get overwhelmed sometimes, so no chances here.